operator-framework-core/src/test/java/io/javaoperatorsdk/operator/LeaderElectionManagerTest.java

```diff
@@ -0,0 +1,75 @@
+package io.javaoperatorsdk.operator;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.javaoperatorsdk.operator.api.config.LeaderElectionConfiguration;
+
+import static io.fabric8.kubernetes.client.Config.KUBERNETES_AUTH_TRYKUBECONFIG_SYSTEM_PROPERTY;
+import static io.fabric8.kubernetes.client.Config.KUBERNETES_NAMESPACE_FILE;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.mock;
+
+class LeaderElectionManagerTest {
+
+  private ControllerManager controllerManager;
+  private KubernetesClient kubernetesClient;
+  private LeaderElectionManager leaderElectionManager;
+
+  @BeforeEach
+  void setUp() {
+    controllerManager = mock(ControllerManager.class);
+    kubernetesClient = mock(KubernetesClient.class);
+    leaderElectionManager = new LeaderElectionManager(controllerManager);
+  }
+
+  @AfterEach
+  void tearDown() {
+    System.getProperties().remove(KUBERNETES_NAMESPACE_FILE);
+    System.getProperties().remove(KUBERNETES_AUTH_TRYKUBECONFIG_SYSTEM_PROPERTY);
+  }
+
+  @Test
+  void testInit() {
+    leaderElectionManager.init(new LeaderElectionConfiguration("test", "testns"), kubernetesClient);
+    assertTrue(leaderElectionManager.isLeaderElectionEnabled());
+  }
+
+  @Test
+  void testInitInferLeaseNamespace(@TempDir Path tempDir) throws IOException {
+    var namespace = "foo";
+    var namespacePath = tempDir.resolve("namespace");
+    Files.writeString(namespacePath, namespace);
+
+    System.setProperty(KUBERNETES_AUTH_TRYKUBECONFIG_SYSTEM_PROPERTY, "false");
```

Comment:
It is because of https://github.com/fabric8io/kubernetes-client/blob/0742e942eb09f8ed2672f7f82e176b76e595ad50/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/Config.java#L819. Otherwise it will try to read `.kube/config`. By setting this the test case makes sure it tries to read the file.

Reply:
I'm however still not sure this gets the correct client config.

Reply:
The order also seems to be wrong:

```java
configurationService.getLeaderElectionConfiguration()
    .ifPresent(c -> leaderElectionManager.init(c, this.kubernetesClient));
ConfigurationServiceProvider.set(configurationService);
```

Now the PR as it is, tries to access `ConfigurationServiceProvider.instance()` before `ConfigurationServiceProvider.set(configurationService)` is called.

For the use case creating k8s client manually, this seems to be required but maybe it is supposed to be so?

```java
var operator =
        new Operator(
            kubernetesClient,
            configurationServiceOverrider ->
                configurationServiceOverrider
                    .withCloseClientOnStop(false)
                    .withClientConfiguration(kubernetesClient.getConfiguration()));
```

Reply:
Hmm, the instance is already set in `ConfigurationServiceProvider::overrideCurrent`, so `ConfigurationServiceProvider.set(configurationService);` doesn't do anything.

Reply:
I have a feeling coupling reading namespace from inside a pod with client config seems to complicate things.

Reply:
Will this be simpler?

```java
final var leaseNamespace =
        config.getLeaseNamespace().orElseGet(
            () -> client.getConfiguration().getNamespace());
```

Reply:
```
final var leaseNamespace =
        config.getLeaseNamespace().orElseGet(
            () -> client.getConfiguration().getNamespace());
```
This sounds reasonable for me, also it is good that there is a null check after. Since client namespace can be null too.

Reply:
It might be good to do an Integration test.

Reply:
To me there seems to be two separated issues:

1. In case of manually created k8s client, it is better to enforce setting `withClientConfiguration` otherwise there could be a mismatch
2. Which is considered to be a better way retrieving configuration, from client directly or from `ConfigurationServiceProvider`?

I can try to add an integration test.

Reply:
> In case of manually created k8s client, it is better to enforce setting withClientConfiguration otherwise there could be a mismatch

What do you mean manually created?  In the background it always created, if the namespace is not set it should use always one where the pod is running.

> Which is considered to be a better way retrieving configuration, from client directly or from ConfigurationServiceProvider?

I would say the config service provider since that can be overridden. But I don't really understand that method, shouldn't that call the client in the implementation?
` operator.getKubernetesClient().getConfiguration() `. @metacosm  ? 

So the client should be part of configuration service rather. WDYT?

Reply:
That might a separate issue to improve this config part probably, I would just use the config service for the sake of this issue.

Reply:
> What do you mean manually created? In the background it always created, if the namespace is not set it should use always one where the pod is running.

For example:

```java
var kubernetesClient = ...; # build the client; may not be from .kube/config
var operator =
        new Operator(
            kubernetesClient,
            configurationServiceOverrider ->
                configurationServiceOverrider
                    .withCloseClientOnStop(false)
                    .withClientConfiguration(kubernetesClient.getConfiguration()));
```

If here `withClientConfiguration` is not called, I think a default configuration will be created, which might be different than the configuration used to build `kubernetesClient`. Then we have an inconsistency. I'm not sure of the consequence though.

Reply:
Yes, that is absolutely true. The thing is that on the other hand on quarkus extension it would make issues when the config from config service not used. Note however that if the namespace is configured for a client, it's not our api.  
So we have api for namespace config on controller level. And we have api for leader election to config namespace. So I would expect that the leader election is either uses the pod namespace or a setting explicitly for it.

Reply:
Yeah I understand that. I think as you said this is a separated issue to track, not related to leader election. I can create a new issue.

Reply:
I modified the existing e2e test a bit. PTAL, thanks.

Reply:
We probably need to clean things up or, at the very least, document them better. The intention is to create client instance(s) based on the configuration retrieved from the `ConfigurationService`. I'm not sure that we do follow that ourselves in our code but we should review it. You should only create client instances with a different configuration only for very specific purposes.